File: src/protocol/CommandFactory.cpp
#include "CommandFactory.h"
#include "BlindsDOWNOnTimeBuilder.h"
#include "BlindsUPBuilder.h"
#include "BlindsStatusBuilder.h"
#include "BlindsDOWNBuilder.h"
#include "BlindsUPOnTimeBuilder.h"
#include "AutomaticBlindsBuilder.h"

#include <stdexcept>

using namespace std;
using namespace communication;

CommandFactory::CommandFactory()
{}

CommandFactory::~CommandFactory()
{}

unique_ptr<Command> CommandFactory::createCommand(const vector<uint8_t> &commandInBytes)
{
    const auto type = static_cast<CommandType>(commandInBytes[COMMAND_TYPE_POSITION]);

    switch (type)
    {
        case CommandType::BLINDS_UP :
            builder_ = make_unique<BlindsUPBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::BLINDS_DOWN :
            builder_ = make_unique<BlindsDOWNBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::BLINDS_UP_ON_TIME :
            builder_ = make_unique<BlindsUPOnTimeBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::BLINDS_DOWN_ON_TIME :
            builder_ = make_unique<BlindsDOWNOnTimeBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::BLINDS_STATUS :
            builder_ = make_unique<BlindsStatusBuilder>();
            return move(builder_->create(commandInBytes));
        case CommandType::AUTOMATIC_BLINDS :
            builder_ = make_unique<AutomaticBlindsBuilder>();
            return move(builder_->create(commandInBytes));
        default:
            throw invalid_argument("Received command does not register in factory.");

    }
}

File: mobile_client/app/src/main/java/protocol/CommandFactory.java
package protocol;
import java.io.*;
import java.util.ArrayList;

public class CommandFactory implements Closeable
	{
		private CommandBuilder builder;
		private final int COMMAND_TYPE_POSITION = 5;

		public CommandFactory()
		{
		}
		public final void close()
		{
		}

		public Command createCommand(ArrayList<Integer> commandInBytes)
		{
			CommandType type = CommandType.forValue(commandInBytes.get(COMMAND_TYPE_POSITION));
			Command command;
			switch (type)
			{
				case BLINDS_DOWN:
				{
					this.builder = new BlindsDOWNBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case BLINDS_DOWN_ON_TIME:
				{
					this.builder = new BlindsDOWNOnTimeBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case BLINDS_UP:
				{
					this.builder = new BlindsUPBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case BLINDS_UP_ON_TIME:
				{
					this.builder = new BlindsUPOnTimeBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case SNAPSHOT:
				{
					this.builder = new SnapshotBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case AUTOMATIC_BLINDS:
				{
					this.builder = new AutomaticBlindsBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case BLINDS_STATUS:
				{
					this.builder = new BlindsStatusBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case USER_OUT_OF_HOME:
				{
					this.builder = new UserOutOfHomeBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case TEMPERATURE_DEMAND:
				{
					this.builder = new TemperatureDemandBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				case END_CONNECTION:
				{
					this.builder = new EndConnectionBuilder();
					command = this.builder.create(commandInBytes);
					break;
				}
				default:
				{
					throw new IllegalArgumentException("Score is not valid!");
				}
			}
			return command;
		}
	}